From a6ac84731776381ee5ac6b6306acdc4418e900d9 Mon Sep 17 00:00:00 2001
From: Tim Blechmann <tim.blechmann@qt.io>
Date: Thu, 22 May 2025 10:46:09 +0200
Subject: [PATCH 479/553] Long live Q_PRESUME

Q_PRESUME wraps a Q_ASSERT/[[assume]] and can be used as a stop-gap
until [[assume]] can be used unconditionally. It has stricter
semantics than the deprecated Q_ASSUME and is helpful to silence static
analyzer warnings.

Documentation and [ChangeLog] in the next commit so it won't be
cherry-picked.

Fixes: QTBUG-141074
Change-Id: Id5376bcc5e9e9708c836ceff5eea982c2b0e382e
Reviewed-by: Giuseppe D'Angelo <giuseppe.dangelo@kdab.com>
Reviewed-by: Tim Blechmann <tim.blechmann@qt.io>
Reviewed-by: Fabian Kosmale <fabian.kosmale@qt.io>
(cherry picked from commit 0a0aed0fcd2eeeb28071ccc1ab1b04a781076133)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/global/qassert.h                  | 19 +++++++++++++++++++
 src/corelib/text/qbytearray.h                 |  4 +---
 src/corelib/text/qstring.h                    |  4 +---
 src/corelib/tools/qlist.h                     |  4 +---
 .../corelib/global/qglobal/tst_qglobal.cpp    | 13 +++++++++++++
 5 files changed, 35 insertions(+), 9 deletions(-)

diff --git a/src/corelib/global/qassert.h b/src/corelib/global/qassert.h
index d1d306fd3ed..05210acb2d4 100644
--- a/src/corelib/global/qassert.h
+++ b/src/corelib/global/qassert.h
@@ -100,6 +100,25 @@ inline bool qt_assume_is_deprecated(bool cond) noexcept { return cond; }
         Q_ASSUME_IMPL(valueOfExpression);\
     }(qt_assume_is_deprecated(Expr))
 
+
+#if __has_builtin(__builtin_assume)
+// Clang has this intrinsic and won't warn about its use in C++20 mode
+#  define Q_PRESUME_IMPL(assumption) __builtin_assume(assumption)
+#elif __has_cpp_attribute(assume)
+// GCC has implemented this attribute and allows its use in C++20 mode
+#  define Q_PRESUME_IMPL(assumption) [[assume(assumption)]]
+#elif defined(Q_CC_MSVC)
+#  define Q_PRESUME_IMPL(assumption) __assume(assumption)
+#else
+#  define Q_PRESUME_IMPL(assumption) (void)0
+#endif
+
+#define Q_PRESUME(assumption)       \
+    [&] {                            \
+        Q_ASSERT(assumption);       \
+        Q_PRESUME_IMPL(assumption); \
+    }()
+
 // Don't use these in C++ mode, use static_assert directly.
 // These are here only to keep old code compiling.
 #  define Q_STATIC_ASSERT(Condition) static_assert(bool(Condition), #Condition)
diff --git a/src/corelib/text/qbytearray.h b/src/corelib/text/qbytearray.h
index 49d9e24d036..545d7d39f6b 100644
--- a/src/corelib/text/qbytearray.h
+++ b/src/corelib/text/qbytearray.h
@@ -512,10 +512,8 @@ public:
     }
     constexpr qsizetype size() const noexcept
     {
-#if __has_cpp_attribute(assume)
         constexpr size_t MaxSize = maxSize();
-        [[assume(size_t(d.size) <= MaxSize)]];
-#endif
+        Q_PRESUME(size_t(d.size) <= MaxSize);
         return d.size;
     }
 #if QT_DEPRECATED_SINCE(6, 4)
diff --git a/src/corelib/text/qstring.h b/src/corelib/text/qstring.h
index 2e4a92cb8d2..1c8be2ea167 100644
--- a/src/corelib/text/qstring.h
+++ b/src/corelib/text/qstring.h
@@ -237,10 +237,8 @@ public:
     }
     constexpr qsizetype size() const noexcept
     {
-#if __has_cpp_attribute(assume)
         constexpr size_t MaxSize = maxSize();
-        [[assume(size_t(d.size) <= MaxSize)]];
-#endif
+        Q_PRESUME(size_t(d.size) <= MaxSize);
         return d.size;
     }
 #if QT_DEPRECATED_SINCE(6, 4)
diff --git a/src/corelib/tools/qlist.h b/src/corelib/tools/qlist.h
index da0d89c021a..fb28866e81a 100644
--- a/src/corelib/tools/qlist.h
+++ b/src/corelib/tools/qlist.h
@@ -446,10 +446,8 @@ public:
     static constexpr qsizetype maxSize() { return Data::maxSize(); }
     constexpr qsizetype size() const noexcept
     {
-#if __has_cpp_attribute(assume)
         constexpr size_t MaxSize = maxSize();
-        [[assume(size_t(d.size) <= MaxSize)]];
-#endif
+        Q_PRESUME(size_t(d.size) <= MaxSize);
         return d.size;
     }
     constexpr qsizetype count() const noexcept { return size(); }
diff --git a/tests/auto/corelib/global/qglobal/tst_qglobal.cpp b/tests/auto/corelib/global/qglobal/tst_qglobal.cpp
index 3804f53cae9..0b65673c393 100644
--- a/tests/auto/corelib/global/qglobal/tst_qglobal.cpp
+++ b/tests/auto/corelib/global/qglobal/tst_qglobal.cpp
@@ -92,6 +92,7 @@ private slots:
     void qIsNull();
     void for_each();
     void qassert();
+    void qpresume();
     void qtry();
     void checkptr();
     void qstaticassert();
@@ -266,6 +267,18 @@ void tst_QGlobal::qassert()
     QVERIFY(passed);
 }
 
+/*non-static*/ Q_NEVER_INLINE char presumedValue(const char *str)
+{
+    Q_PRESUME(str);
+    // an optimizing compiler should delete the str? check
+    return str ? str[0] : '\0';
+}
+
+void tst_QGlobal::qpresume()
+{
+    QCOMPARE(presumedValue("Hello World"), 'H');
+}
+
 void tst_QGlobal::qtry()
 {
     int i = 0;
-- 
2.51.2

